localization/yabloc/yabloc_imgproc/src/segment_filter/segment_filter_core.cpp

```diff
@@ -0,0 +1,282 @@
+{
+  float min_distance = std::min(pn.x, pn.normal_x);
+  float max_distance = std::max(pn.x, pn.normal_x);
+  if (min_distance > max_segment_distance_) return false;
+  if (max_distance < max_segment_distance_) {
+    truncated_pn = pn;
+    return true;
+  }
+
+  truncated_pn = pn;
+  Eigen::Vector3f t = pn.getVector3fMap() - pn.getNormalVector3fMap();
+  float not_zero_tx = t.x() > 0 ? std::max(t.x(), 1e-3f) : std::min(t.x(), -1e-3f);
+  float lambda = (max_segment_distance_ - pn.x) / not_zero_tx;
+  Eigen::Vector3f m = pn.getVector3fMap() + lambda * t;
+  if (pn.x > pn.normal_x)
+    truncated_pn.getVector3fMap() = m;
+  else
+    truncated_pn.getNormalVector3fMap() = m;
+  return true;
+}
+
+std::set<ushort> get_unique_pixel_value(cv::Mat & image)
+{
+  // `image` is a set of ushort.
+  // The purpose is to find the unduplicated set of values contained in `image`.
+  // For example, if `image` is {0,1,2,0,1,2,3}, this function returns {0,1,2,3}.
+
+  if (image.depth() != CV_16U) throw std::runtime_error("image's depth must be ushort");
+
+  auto begin = image.begin<ushort>();
+  auto last = std::unique(begin, image.end<ushort>());
+  std::sort(begin, last);
+  last = std::unique(begin, last);
+  return std::set<ushort>(begin, last);
+}
+
+pcl::PointCloud<pcl::PointNormal> SegmentFilter::project_lines(
+  const pcl::PointCloud<pcl::PointNormal> & points, const std::set<int> & indices,
+  bool negative) const
+{
+  pcl::PointCloud<pcl::PointNormal> projected_points;
+  for (size_t index = 0; index < points.size(); ++index) {
+    if (negative) {
+      if (indices.count(index) > 0) continue;
+    } else {
+      if (indices.count(index) == 0) continue;
+    }
+
+    pcl::PointNormal truncated_pn = points.at(index);
+
+    std::optional<Eigen::Vector3f> opt1 = project_func_(truncated_pn.getVector3fMap());
+    std::optional<Eigen::Vector3f> opt2 = project_func_(truncated_pn.getNormalVector3fMap());
+    if (!opt1.has_value()) continue;
+    if (!opt2.has_value()) continue;
+
+    // If linesegment has shoter length than config, it is excluded
+    if (min_segment_length_ > 0) {
+      float length = (opt1.value() - opt2.value()).norm();
+      if (length < min_segment_length_) continue;
+    }
+    if (max_lateral_distance_ > 0) {
+      float abs_lateral1 = std::abs(opt1.value().y());
+      float abs_lateral2 = std::abs(opt2.value().y());
+      if (std::min(abs_lateral1, abs_lateral2) > max_lateral_distance_) continue;
+    }
+
+    pcl::PointNormal xyz;
+    xyz.x = opt1->x();
+    xyz.y = opt1->y();
+    xyz.z = opt1->z();
+    xyz.normal_x = opt2->x();
+    xyz.normal_y = opt2->y();
+    xyz.normal_z = opt2->z();
+
+    //
+    pcl::PointNormal truncated_xyz = xyz;
+    if (max_segment_distance_ > 0)
+      if (!is_near_element(xyz, truncated_xyz)) continue;
+
+    projected_points.push_back(truncated_xyz);
+  }
+  return projected_points;
+}
+
+std::set<int> SegmentFilter::filt_by_mask(
```

Comment:
`filt_by_mask` -> `filter_by_mask`

Reply:
It sounds nice! Your review is reflected in d9617f3. :+1: